### PR TITLE
Feat/met 6467 abstract engine connection

### PR DESCRIPTION
**Pull request by stzanakis**

### CODE REVIEW

File: metis-core-common/src/main/java/eu/europeana/metis/core/engine/base/OaiHarvestParameters.java [hunk 1]
```diff
@@ -0,0 +1,11 @@
+package eu.europeana.metis.core.engine.base;
+
+import java.util.Date;
+
```
Comment (jeortizquan) [inline metis-core-common/src/main/java/eu/europeana/metis/core/engine/base/OaiHarvestParameters.java]: remember to add java doc to all new files.

File: metis-core-common/src/main/java/eu/europeana/metis/core/engine/base/ProcessingEngineTaskClient.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+/**
+ * Client for submitting, monitoring and cancelling external tasks.
+ */
+public interface ProcessingEngineTaskClient<S extends ProcessingEngineTaskSettings, T extends ProcessingEngineTask> {
+
+  S getProcessingEngineTaskSettings();
+
+  Supplier<T> getTaskCreator();
+
+  long submitTask(T externalTask, String topologyName) throws ExternalTaskException;
```
Comment (jeortizquan) [inline metis-core-common/src/main/java/eu/europeana/metis/core/engine/base/ProcessingEngineTaskClient.java]: suggestion, if we don't want the variables be reassigned I would add `final` keyword to all the parameters where applicable, saying this comment out of the perspective of security, and immutability.

```
 long submitTask(final T externalTask, final String topologyName) throws ExternalTaskException;

ProcessingEngineTaskProgress getTaskProgress(final String topologyName, final long taskId)
      throws ExternalTaskException, UnrecoverableExternalTaskException;

  long getTotalIndexedRecords(final String datasetId, final IndexDatabase indexDatabase) throws ExternalTaskException;

  List<String> getPublishedRecords(final String datasetId, final List<String> recordsIds) throws ExternalTaskException;

  Map<String, Boolean> getRecordStatus(final String topologyName, final long taskId, final int from, final int to) throws ExternalTaskException;

  List<DataItemStatus> getExternalRecordStatuses(final String topologyName, final long taskId, final int from, final int to) throws ExternalTaskException;
 void cancel(final String topologyName, final long taskId, final String message) throws ExternalTaskException;
```

File: metis-core-common/src/main/java/eu/europeana/metis/core/engine/base/ProcessingEngineTaskKeys.java [hunk 3]
```diff
@@ -0,0 +1,33 @@
+package eu.europeana.metis.core.engine.base;
+
+public enum ProcessingEngineTaskKeys {
```
Comment (jeortizquan) [inline metis-core-common/src/main/java/eu/europeana/metis/core/engine/base/ProcessingEngineTaskKeys.java]: java doc

File: metis-core-common/src/main/java/eu/europeana/metis/core/engine/base/ProcessingEngineTaskConfigurator.java [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+    parameters.put(NEW_REPRESENTATION_NAME, MetisPlugin.getRepresentationName());
+    parameters.put(OUTPUT_DATA_SETS, String
+        .format(CommonStringValues.S_DATA_PROVIDERS_S_DATA_SETS_S_TEMPLATE, processingEngineTaskSettings.baseUrl(),
+            processingEngineTaskSettings.provider(), externalDatasetId));
+    return parameters;
+  }
+
+  public static <S extends ProcessingEngineTaskSettings, T extends ProcessingEngineTask>
+  Map<ProcessingEngineTaskKeys, String> createDefaultTaskParametersHarvest(
+      String datasetId, boolean incrementalHarvest, Date startedDate,
```
Comment (jeortizquan) [inline metis-core-common/src/main/java/eu/europeana/metis/core/engine/base/ProcessingEngineTaskConfigurator.java]: my suggestion is that this parameter `boolean incrementalHarvest` should be changed to an e.g.  `enum harvestType { FULL, INCREMENTAL}`

File: metis-core-common/src/main/java/eu/europeana/metis/core/engine/base/ProcessingEngineTaskConfigurator.java [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+    if (performSampling && sampleSize != null) {
+      parameters.put(SAMPLE_SIZE, sampleSize.toString());
+    }
+    return parameters;
+  }
+
+  public static Map<ProcessingEngineTaskKeys, String> createParametersIndex(
+      String datasetId,
+      Date pluginStartedDate,
+      boolean incrementalIndexing,
```
Comment (jeortizquan) [inline metis-core-common/src/main/java/eu/europeana/metis/core/engine/base/ProcessingEngineTaskConfigurator.java]: in accordance to the previous suggestion, also change here to an `enum indexType {FULL, INCREMENTAL}`

File: metis-core-common/src/main/java/eu/europeana/metis/core/engine/base/report/item/DataItemState.java [hunk 6]
```diff
@@ -0,0 +1,9 @@
+package eu.europeana.metis.core.engine.base.report.item;
+
```
Comment (jeortizquan) [inline metis-core-common/src/main/java/eu/europeana/metis/core/engine/base/report/item/DataItemState.java]: javadoc

File: metis-core-common/src/main/java/eu/europeana/metis/core/engine/base/ProcessingEngineTaskConfigurator.java [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+import java.util.Date;
+import java.util.EnumMap;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Set;
+import java.util.TimeZone;
+import org.springframework.util.CollectionUtils;
+
```
Comment (jeortizquan) [inline metis-core-common/src/main/java/eu/europeana/metis/core/engine/base/ProcessingEngineTaskConfigurator.java]: javadoc

File: metis-core-common/src/main/java/eu/europeana/metis/core/engine/base/ProcessingEngineTask.java [hunk 8]
```diff
@@ -0,0 +1,19 @@
+package eu.europeana.metis.core.engine.base;
+
+import java.util.Map;
+
+/**
+ * Represents a task that can be submitted to an external system.
+ */
+public interface ProcessingEngineTask {
+
```
Comment (jeortizquan) [inline metis-core-common/src/main/java/eu/europeana/metis/core/engine/base/ProcessingEngineTask.java]: suggestion consider if `final` keyword also applies here to all parameters.

File: metis-core-common/src/main/java/eu/europeana/metis/core/engine/base/report/item/DataItemStatus.java [hunk 9]
```diff
@@ -0,0 +1,11 @@
+package eu.europeana.metis.core.engine.base.report.item;
+
```
Comment (jeortizquan) [inline metis-core-common/src/main/java/eu/europeana/metis/core/engine/base/report/item/DataItemStatus.java]: javadoc

File: metis-core-common/src/main/java/eu/europeana/metis/core/engine/base/report/item/content/ContentAttributeStatistics.java [hunk 10]
```diff
@@ -0,0 +1,4 @@
+package eu.europeana.metis.core.engine.base.report.item.content;
+
```
Comment (jeortizquan) [inline metis-core-common/src/main/java/eu/europeana/metis/core/engine/base/report/item/content/ContentAttributeStatistics.java]: javadoc

File: metis-core-common/src/main/java/eu/europeana/metis/core/engine/base/report/item/content/ContentNodeReport.java [hunk 11]
```diff
@@ -0,0 +1,7 @@
+package eu.europeana.metis.core.engine.base.report.item.content;
+
+import java.util.List;
+
```
Comment (jeortizquan) [inline metis-core-common/src/main/java/eu/europeana/metis/core/engine/base/report/item/content/ContentNodeReport.java]: javadoc

File: metis-core-common/src/main/java/eu/europeana/metis/core/engine/base/report/task/ProcessingEngineTaskState.java [hunk 12]
```diff
@@ -0,0 +1,27 @@
+package eu.europeana.metis.core.engine.base.report.task;
+
```
Comment (jeortizquan) [inline metis-core-common/src/main/java/eu/europeana/metis/core/engine/base/report/task/ProcessingEngineTaskState.java]: javadoc

File: metis-core-common/src/main/java/eu/europeana/metis/core/engine/base/report/item/content/ContentNodeStatistics.java [hunk 13]
```diff
@@ -0,0 +1,7 @@
+package eu.europeana.metis.core.engine.base.report.item.content;
+
+import java.util.Set;
+
```
Comment (jeortizquan) [inline metis-core-common/src/main/java/eu/europeana/metis/core/engine/base/report/item/content/ContentNodeStatistics.java]: javadoc

File: metis-core-common/src/main/java/eu/europeana/metis/core/engine/base/report/item/content/ContentStatisticsReport.java [hunk 14]
```diff
@@ -0,0 +1,6 @@
+package eu.europeana.metis.core.engine.base.report.item.content;
+
+import java.util.List;
+
```
Comment (jeortizquan) [inline metis-core-common/src/main/java/eu/europeana/metis/core/engine/base/report/item/content/ContentStatisticsReport.java]: javadoc.

File: metis-core-common/src/main/java/eu/europeana/metis/core/engine/base/report/task/ProcessingEngineTaskErrorDetails.java [hunk 15]
```diff
@@ -0,0 +1,4 @@
+package eu.europeana.metis.core.engine.base.report.task;
+
```
Comment (jeortizquan) [inline metis-core-common/src/main/java/eu/europeana/metis/core/engine/base/report/task/ProcessingEngineTaskErrorDetails.java]: javadoc

File: metis-core-common/src/main/java/eu/europeana/metis/core/engine/base/report/task/ProcessingEngineTaskErrorInfo.java [hunk 16]
```diff
@@ -0,0 +1,7 @@
+package eu.europeana.metis.core.engine.base.report.task;
+
+import java.util.List;
+
```
Comment (jeortizquan) [inline metis-core-common/src/main/java/eu/europeana/metis/core/engine/base/report/task/ProcessingEngineTaskErrorInfo.java]: javadoc

File: metis-core-common/src/main/java/eu/europeana/metis/core/engine/base/report/task/ProcessingEngineTaskErrors.java [hunk 17]
```diff
@@ -0,0 +1,6 @@
+package eu.europeana.metis.core.engine.base.report.task;
+
+import java.util.List;
+
```
Comment (jeortizquan) [inline metis-core-common/src/main/java/eu/europeana/metis/core/engine/base/report/task/ProcessingEngineTaskErrors.java]: javadoc

File: metis-core-common/src/main/java/eu/europeana/metis/core/engine/ecloud/DpsProcessingEngineTask.java [hunk 18]
```diff
@@ -0,0 +1,60 @@
...
+import eu.europeana.cloud.service.dps.DpsTask;
+import eu.europeana.cloud.service.dps.OAIPMHHarvestingDetails;
+import eu.europeana.metis.core.engine.base.DataRevision;
+import eu.europeana.metis.core.engine.base.ProcessingEngineTask;
+import eu.europeana.metis.core.engine.base.OaiHarvestParameters;
+import eu.europeana.metis.core.engine.base.ProcessingEngineTaskKeys;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
```
Comment (jeortizquan) [inline metis-core-common/src/main/java/eu/europeana/metis/core/engine/ecloud/DpsProcessingEngineTask.java]: javadoc

File: metis-core-common/src/main/java/eu/europeana/metis/core/engine/ecloud/DpsProcessingEngineTaskClient.java [hunk 19]
```diff
@@ -0,0 +1,265 @@
...
+import eu.europeana.metis.exception.ExternalTaskException;
+import eu.europeana.metis.exception.UnrecoverableExternalTaskException;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
```
Comment (jeortizquan) [inline metis-core-common/src/main/java/eu/europeana/metis/core/engine/ecloud/DpsProcessingEngineTaskClient.java]: javadoc

File: metis-core-common/src/main/java/eu/europeana/metis/core/engine/base/EngineRecordClient.java [hunk 20]
```diff
@@ -0,0 +1,78 @@
...
+import eu.europeana.metis.core.rest.Record;
+import eu.europeana.metis.exception.ExternalTaskException;
+import java.util.Date;
+import java.util.List;
+
+/**
+ * Provides methods for retrieving and managing records within the processing engine.
+ */
+public interface EngineRecordClient {
+
```
Comment (jeortizquan) [inline metis-core-common/src/main/java/eu/europeana/metis/core/engine/base/EngineRecordClient.java:12]: my suggestion is to consider `final` in all parameters if values are not intended to change. you will see I made the comment in other places as well.

File: metis-core-common/src/main/java/eu/europeana/metis/core/engine/base/EngineTask.java [hunk 21]
```diff
@@ -0,0 +1,23 @@
...
+import java.util.Objects;
+
+/**
+ * Represents a task that can be submitted to a processing engine.
+ */
+public class EngineTask {
+  protected final Map<EngineTaskKey, String> parameters;
+  protected final InputDataEndpoint inputDataEndpoint;
+  protected final DataRevision outputDataRevision;
+
```
Comment (jeortizquan) [inline metis-core-common/src/main/java/eu/europeana/metis/core/engine/base/EngineTask.java:14]: suggestion also here, see if final applies also here for all parameters for immutability.

File: metis-core-common/src/main/java/eu/europeana/metis/core/engine/base/EngineTaskKey.java [hunk 22]
```diff
@@ -0,0 +1,36 @@
+package eu.europeana.metis.core.engine.base;
+
+/**
+ * Enum representing various keys used to configure and manage tasks within the processing engine.
+ */
+public enum EngineTaskKey {
+  XSLT_URL,
```
Comment (jeortizquan) [inline metis-core-common/src/main/java/eu/europeana/metis/core/engine/base/EngineTaskKey.java:10]: suggestion, javadoc for each specific key?

File: metis-core-common/src/main/java/eu/europeana/metis/core/engine/base/task/input/InputDataEndpoint.java [hunk 23]
```diff
@@ -0,0 +1,17 @@
...
+@SuppressWarnings("javaarchitecture:S7027")
+public sealed interface InputDataEndpoint
+    permits HttpHarvestInputDataEndpoint, OaiHarvestInputDataEndpoint, InternalInputDataEndpoint {
+
+  /**
+   * Retrieves the URL associated with the input data endpoint.
+   *
+   * @return The URL of the input data endpoint.
+   */
+  String url();
```
Comment (jeortizquan) [inline metis-core-common/src/main/java/eu/europeana/metis/core/engine/base/task/input/InputDataEndpoint.java:16]: suggestion to make it little more verbose. `inputEndpointUrl` or `inputUrl`

File: metis-core-common/src/main/java/eu/europeana/metis/core/engine/ecloud/EcloudEngineRecordStatisticsConverter.java [hunk 24]
```diff
@@ -0,0 +1,91 @@
...
+                    .sorted(Comparator.comparing(NodePathStatisticsDTO::xPath)).toList();
+    return new RecordStatisticsDTO(Long.toString(report.getTaskId()), nodePathStatisticsDTOList);
+  }
+
+  /**
+   * Compiles statistics for a specific node path based on the provided node reports.
+   *
+   * @param nodePath The XPath expression representing the node.
+   * @param nodeReports The list of node reports containing statistical data.
+   * @return A NodePath*/
```
Comment (jeortizquan) [inline metis-core-common/src/main/java/eu/europeana/metis/core/engine/ecloud/EcloudEngineRecordStatisticsConverter.java]: javadoc format `*/`

File: metis-core-common/src/main/java/eu/europeana/metis/core/rest/stats/AttributeStatisticsDTO.java [hunk 25]
```diff
@@ -7,4 +7,4 @@
  * @param value The value of the attribute.
  * @param occurrences The number of times this attribute value appears.
  */
-public record AttributeStatistics(String xPath, String value, long occurrences) {}
+public record AttributeStatisticsDTO(String xPath, String value, long occurrences) {}
```
Comment (jeortizquan) [inline metis-core-common/src/main/java/eu/europeana/metis/core/rest/stats/AttributeStatisticsDTO.java:10]: suggestion, would it be better to name a package `eu.europeana.metis.core.rest.stats.dto` and everything that is there we know is a DTO and remove the suffix?

File: metis-core-service/src/main/java/eu/europeana/metis/core/execution/PluginExecutor.java [hunk 26]
```diff
@@ -0,0 +1,284 @@
...
+        incrementalHarvest = httpHarvestPluginMetadata.isIncrementalHarvest();
+        targetUrl = httpHarvestPluginMetadata.getUrl();
+      }
+      default -> throw new IllegalStateException("Unexpected value: " + plugin);
+    }
+    return new PluginHarvestParameters(targetUrl, incrementalHarvest, oaiHarvestInputDataParameters);
+  }
+
+  private @NotNull Map<EngineTaskKey, String> getProcessPluginParameters() {
+    return switch (plugin.getPluginMetadata()) {
```
Comment (jeortizquan) [inline metis-core-service/src/main/java/eu/europeana/metis/core/execution/PluginExecutor.java:211]: suggestion for each case I think there is no need of creation of extra variable. You can write something like this and still readable.
```
  case ValidationExternalPluginMetadata validationExternalPluginMetadata -> {
        yield createValidationExternalParameters(
            validationExternalPluginMetadata.getUrlOfSchemasZip(),
            validationExternalPluginMetadata.getSchemaRootPath(),
            validationExternalPluginMetadata.getSchematronRootPath());
      }
```

File: metis-core-common/src/main/java/eu/europeana/metis/core/engine/ecloud/EcloudEngineDatasetRecordClient.java [hunk 27]
```diff
@@ -0,0 +1,224 @@
...
+import org.slf4j.LoggerFactory;
+
+/**
+ * A client for managing interactions with eCloud Engine dataset records, including creation,
+ * retrieval, and verification.
+ */
+public class EcloudEngineDatasetRecordClient {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+  private final DateFormat pluginDateFormatForEcloud = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSSXXX", Locale.US);
```
Comment (jeortizquan) [inline metis-core-common/src/main/java/eu/europeana/metis/core/engine/ecloud/EcloudEngineDatasetRecordClient.java:41]: suggestion, to consider in the future. see something like `Locale.EU_EN` to switch in all the codebase.

File: metis-core-common/src/main/java/eu/europeana/metis/core/engine/ecloud/EcloudEngineTask.java [hunk 28]
```diff
@@ -0,0 +1,77 @@
...
+
+/**
+ * Represents a task for the Ecloud processing engine that wraps and transforms
+ * task parameters for use in a DPS task.
+ */
+public class EcloudEngineTask extends EngineTask {
+
+  private final DpsTask dpsTask = new DpsTask();
+
+  public EcloudEngineTask(Map<EngineTaskKey, String> parameters, InputDataEndpoint inputDataEndpoint,
```
Comment (jeortizquan) [inline metis-core-common/src/main/java/eu/europeana/metis/core/engine/ecloud/EcloudEngineTask.java:36]: javadoc of constructor

File: metis-core-common/src/main/java/eu/europeana/metis/core/engine/ecloud/EcloudEngineTask.java [hunk 29]
```diff
@@ -0,0 +1,77 @@
...
+  public DpsTask toDpsTask() {
+    return dpsTask;
+  }
+
+  private void setParameters() {
+    parameters.forEach((key, value) -> dpsTask.addParameter(key.name(), value));
+  }
+
+  private void setInputDataLocation() {
+    final InputDataType inputDataType = switch (inputDataEndpoint) {
```
Comment (jeortizquan) [inline metis-core-common/src/main/java/eu/europeana/metis/core/engine/ecloud/EcloudEngineTask.java]: suggestion the attribute seems obfuscated when reading the code maybe write it as `this.inputDataEndpoint` or add parameter `private void setupInputDataLocation(InputDataEndpoint inputDataEndpoint)`

File: metis-core-common/src/main/java/eu/europeana/metis/core/workflow/plugins/ExecutablePluginType.java [hunk 30]
```diff
@@ -80,4 +79,17 @@ public static ExecutablePluginType getPluginTypeFromEnumName(
...
     return null;
   }
+
+  public ExecutablePluginTypeGroup getExecutablePluginTypeGroup() {
+    return executablePluginTypeGroup;
+  }
+
+  /**
+   * Enum representing groups of executable plugin types.
+   * <p>
```
Comment (jeortizquan) [inline metis-core-common/src/main/java/eu/europeana/metis/core/workflow/plugins/ExecutablePluginType.java:89]: unclosed tag

File: metis-core-rest/src/main/java/eu/europeana/metis/core/rest/config/EngineClientConfig.java [hunk 31]
```diff
@@ -0,0 +1,162 @@
...
+        metisCoreConfigurationProperties.dpsConnectTimeoutInMilliseconds(),
+        metisCoreConfigurationProperties.dpsReadTimeoutInMilliseconds());
+  }
+
+  /**
+   * Closes all initialized service clients, releasing their resources to prevent potential memory leaks. If a client is null, no
+   * action is performed for that client.
+   */
+  @PreDestroy
+  public void close() {
```
Comment (jeortizquan) [inline metis-core-rest/src/main/java/eu/europeana/metis/core/rest/config/EngineClientConfig.java:145]: now that I see this. suggestion or thought? those the order of call here matters?
from this 
```
    if (dpsClient != null) {
      dpsClient.close();
    }
    if (dataSetServiceClient != null) {
      dataSetServiceClient.close();
    }
    if (recordServiceClient != null) {
      recordServiceClient.close();
    }
    if (fileServiceClient != null) {
      fileServiceClient.close();
    }
    if (uisClient != null) {
      uisClient.close();
    }
```
to this
```    
    if (dataSetServiceClient != null) {
      dataSetServiceClient.close();
    }
    if (recordServiceClient != null) {
      recordServiceClient.close();
    }
    if (fileServiceClient != null) {
      fileServiceClient.close();
    }
    if (dpsClient != null) {
      dpsClient.close();
    }
    if (uisClient != null) {
      uisClient.close();
    }
```

File: metis-core-service/src/test/java/eu/europeana/metis/core/execution/TestPluginExecutor.java [hunk 32]
```diff
@@ -0,0 +1,340 @@
...
+  @Mock
+  private EngineTaskSettings engineTaskSettings;
+  private static final String TASK_ID = "taskId";
+  private static final String DATASET_ID = "datasetId";
+  private static final String PREVIOUS_TASK_ID = "previousTaskId";
+  private static final String HARVEST_URL = "http://harvest.url";
+
+  @BeforeEach
+  void setUp() {
+    lenient().when(engineTaskClient.getEngineTaskSettings()).thenReturn(engineTaskSettings);
```
Comment (jeortizquan) [inline metis-core-service/src/test/java/eu/europeana/metis/core/execution/TestPluginExecutor.java:94]: why was needed `lenient()`?  was it because of duplicate stubbing and you it was throwing`fail with 'unnecessary stubbing'` or was something else?

File: metis-core-service/src/main/java/eu/europeana/metis/core/service/ProxiesService.java [hunk 33]
```diff
@@ -34,56 +28,41 @@
...
+import java.lang.invoke.MethodHandles;
 import java.util.Collections;
 import java.util.List;
-import java.util.Locale;
-import java.util.Optional;
-import org.apache.commons.io.IOUtils;
 import org.apache.commons.lang3.tuple.ImmutablePair;
 import org.apache.commons.lang3.tuple.Pair;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
```
Comment (jeortizquan) [inline metis-core-service/src/main/java/eu/europeana/metis/core/service/ProxiesService.java:38]: imports order.

**Review by jeortizquan — COMMENTED** [review]
There are some small suggestions and comments, consider the ones that may apply in this iteration. Overall there is a new thought/design, if possible to add `final` to all parameters of the function to promote immutability of values, this will pave the way to the distributed concurrency of flink. There is one sonar issue that can be addressed `Record record` to `Record recordSomething`.

File: metis-core-common/src/main/java/eu/europeana/metis/core/engine/base/OaiHarvestParameters.java (hunk 1 above)
Comment (stzanakis) [reply metis-core-common/src/main/java/eu/europeana/metis/core/engine/base/OaiHarvestParameters.java]: This file does not exist, same goes for many outdated comments.

**Comment by stzanakis** [comment]
@jeortizquan Please re-check your review.
Many comments are outdated, I assume you made them on an earlier commit.

File: metis-core-common/src/main/java/eu/europeana/metis/core/engine/base/EngineRecordClient.java (hunk 20 above)
Comment (stzanakis) [reply metis-core-common/src/main/java/eu/europeana/metis/core/engine/base/EngineRecordClient.java:12]: We are not doing that generally though. Oh and adding it to the interface it does not mean it is final on the subclass.
Happy to add them but perhaps it bring too much clutter, what do you think?
This applies to the other similar comments.

File: metis-core-common/src/main/java/eu/europeana/metis/core/engine/base/task/input/InputDataEndpoint.java (hunk 23 above)
Comment (stzanakis) [reply metis-core-common/src/main/java/eu/europeana/metis/core/engine/base/task/input/InputDataEndpoint.java:16]: so you have:
InputDataEndpoint inputdataEndpoint;
inputDataEndpoint.url()
OR
inputDataEndpoint.inputEndpointUrl()

Isn't the first one better?

File: metis-core-common/src/main/java/eu/europeana/metis/core/engine/ecloud/EcloudEngineRecordStatisticsConverter.java (hunk 24 above)
Comment (stzanakis) [reply metis-core-common/src/main/java/eu/europeana/metis/core/engine/ecloud/EcloudEngineRecordStatisticsConverter.java]: done

File: metis-core-common/src/main/java/eu/europeana/metis/core/rest/stats/AttributeStatisticsDTO.java (hunk 25 above)
Comment (stzanakis) [reply metis-core-common/src/main/java/eu/europeana/metis/core/rest/stats/AttributeStatisticsDTO.java:10]: I would say not, because in code that is not clearly visible.

File: metis-core-service/src/main/java/eu/europeana/metis/core/execution/PluginExecutor.java (hunk 26 above)
Comment (stzanakis) [reply metis-core-service/src/main/java/eu/europeana/metis/core/execution/PluginExecutor.java:211]: Of course. I thought it was clear assigning the value first. (I assume this applies to the other ones as well?)
Happy to change, what do you prefer?

File: metis-core-common/src/main/java/eu/europeana/metis/core/engine/base/EngineTask.java (hunk 21 above)
Comment (stzanakis) [reply metis-core-common/src/main/java/eu/europeana/metis/core/engine/base/EngineTask.java:14]: done

File: metis-core-common/src/main/java/eu/europeana/metis/core/engine/ecloud/EcloudEngineTask.java (hunk 28 above)
Comment (stzanakis) [reply metis-core-common/src/main/java/eu/europeana/metis/core/engine/ecloud/EcloudEngineTask.java:36]: done

File: metis-core-common/src/main/java/eu/europeana/metis/core/engine/ecloud/EcloudEngineTask.java (hunk 29 above)
Comment (stzanakis) [reply metis-core-common/src/main/java/eu/europeana/metis/core/engine/ecloud/EcloudEngineTask.java]: done

File: metis-core-common/src/main/java/eu/europeana/metis/core/workflow/plugins/ExecutablePluginType.java (hunk 30 above)
Comment (stzanakis) [reply metis-core-common/src/main/java/eu/europeana/metis/core/workflow/plugins/ExecutablePluginType.java:89]: This is normal in javadoc for `<p>` and `<br>` to reduce clutter.

File: metis-core-rest/src/main/java/eu/europeana/metis/core/rest/config/EngineClientConfig.java (hunk 31 above)
Comment (stzanakis) [reply metis-core-rest/src/main/java/eu/europeana/metis/core/rest/config/EngineClientConfig.java:145]: No I don't think it matters.
I've converted to your preference.

File: metis-core-service/src/test/java/eu/europeana/metis/core/execution/TestPluginExecutor.java (hunk 32 above)
Comment (stzanakis) [reply metis-core-service/src/test/java/eu/europeana/metis/core/execution/TestPluginExecutor.java:94]: It's because the newer mockito checks the test method and if you don't use the stub it complains.
So I don't need all of the calls in all the test cases therefore I needed to add lenient otherwise I would have to duplicate the stubs everywhere..

File: metis-core-service/src/main/java/eu/europeana/metis/core/service/ProxiesService.java (hunk 33 above)
Comment (stzanakis) [reply metis-core-service/src/main/java/eu/europeana/metis/core/service/ProxiesService.java:38]: Not sure what you mean, they are sorted automatically.

File: metis-core-common/src/main/java/eu/europeana/metis/core/engine/base/EngineTaskKey.java (hunk 22 above)
Comment (stzanakis) [reply metis-core-common/src/main/java/eu/europeana/metis/core/engine/base/EngineTaskKey.java:10]: Ok, I did my best :D
